Docs: refresh "add a new technology" guide, auto-generate model overview, unify the docs build (#787)
NREL/GreenHEART: docs/developer_guide/adding_a_new_technology.md

Reviewer: I generally agree with Kaitlin's comments but think this is very helpful! Thanks John!

In broad strokes, this involves writing performance and cost wrappers for your technology in the format that H2Integrate expects, then adding those to the list of available technologies in the H2Integrate codebase.
We'll first walk through a relatively straightforward example of adding a new technology, then discuss some of the more complex cases you might encounter.

## Choosing baseclasses for your model

Every model in H2Integrate inherits from a small set of baseclasses that wire it
into the rest of the framework. Before writing code, pick the appropriate base
class and configuration class for each piece of your technology:

| Piece               | Baseclass                                                | Config baseclass                |
| ------------------- | -------------------------------------------------------- | ------------------------------- |
| Performance model   | `PerformanceModelBaseClass`                              | `BaseConfig`                    |
| Cost model          | `CostModelBaseClass`                                     | `CostModelBaseConfig`           |

> | Controller (opt.)   | A `PassthroughController` is inserted automatically      | n/a                             |

Reviewer 2: Not sure if we want to point folks towards the controls doc pages for tech level control base classes in the `technology_control_overview.md`. Classes that seem relevant at this point:
- `PyomoStorageControllerBaseClass`
- `StorageOpenLoopControlBase`

Author: Thanks for bringing this up, I've added references to those baseclasses in `class_structure.md` now.

These are defined in `h2integrate/core/model_baseclasses.py` (model baseclasses

> and `CostModelBaseConfig`) and `h2integrate/core/utilities.py` (`BaseConfig`).

Reviewer 2: ```suggestion
General model baseclasses and configs baseclasses are defined in:
- `h2integrate/core/model_baseclasses.py` 
- `h2integrate/core/utilities.py`
```

- **Adding a brand-new technology?** Inherit directly from
  `PerformanceModelBaseClass` and `CostModelBaseClass`, and use `BaseConfig` /

>   `CostModelBaseConfig` for the corresponding configuration classes.

Reviewer 2: ```suggestion
- **Adding a brand-new technology?** Inherit directly from existing baseclasses and configuration baseclasses
       - Performance models use: `PerformanceModelBaseClass` and `BaseConfig`
       - Cost models use: `CostModelBaseClass` and `CostModelBaseConfig`
```

- **Adding a technology that already has a category-specific baseclass?** Inherit
  from that instead. Existing examples include `SolarPerformanceBaseClass`,

>   `WindPerformanceBaseClass`, and `ElectrolyzerPerformanceBaseClass`. These set

Reviewer 2: ```suggestion
- **Adding a technology that already has a category-specific baseclass?** Inherit
  from a category-specific baseclass instead to easily set up shared I/O and commodity attributes for that technology. Existing examples include:
      - `SolarPerformanceBaseClass`,
      - `WindPerformanceBaseClass`
      - `ElectrolyzerPerformanceBaseClass`
```

>   the commodity attributes and any shared I/O for you.

Reviewer 2: ```suggestion

```

```{note}
Category-specific baseclasses are only worth creating when **multiple models
share inputs, outputs, or methods**. The wind module is the canonical example:
both `FlorisWindPlantPerformanceModel` and `PYSAMWindPlantPerformanceModel`
inherit from `WindPerformanceBaseClass` so they share the same wind-resource
discrete input and turbine-rating output. If you are writing a technology model
that doesn't fit into an existing category, skip the intermediate baseclass and
inherit directly from `PerformanceModelBaseClass`.
```

Configuration classes use the [`attrs`](https://www.attrs.org) library and the
`BaseConfig.from_dict` constructor, which validates user-supplied entries from
`tech_config['model_inputs']` against the declared fields. This pattern is now
standard for both performance and cost models in H2Integrate.

## Adding a new technology

We'll walk through the process of adding a solar PV performance model to
H2Integrate. Solar already has a category-specific baseclass, so we'll inherit
from that.

1. **Identify (or write) the relevant baseclass.**
For a solar technology, the category baseclass is
`SolarPerformanceBaseClass`, defined in

> `h2integrate/converters/solar/solar_baseclass.py`:

Reviewer 2: Not a necessary action item for this PR but I think it would be good to unify how baseclass files are named so if you search the files you can easily find all of the existing base classes. Currently we have `baseclass.py`, `baseclasses.py` and `base.py`.

Reviewer 2: #796

        raise NotImplementedError("This method should be implemented in a subclass.")
```

Inheriting from `PerformanceModelBaseClass` (rather than `om.ExplicitComponent`
directly) means the baseclass:

- Declares the standard `driver_config` / `plant_config` / `tech_config` options.
- Reads `n_timesteps`, `dt`, `plant_life`, and `fraction_of_year_simulated` from `plant_config`.
- Validates that `commodity`, `commodity_rate_units`, and `commodity_amount_units` are set on the subclass and registers all of the standard production outputs from those attributes.
- Adds the command-value input and uncurtailed output for `flexible` models, and provides the `apply_curtailment()` helper.

Every performance model must therefore define three class attributes and three commodity attributes; see [Required class attributes](#required-class-attributes) below.

> 2. **Write the performance model for your technology.**

Reviewer 2: Tbh I think it's more relevant to show an example that inherits the `PerformanceBaseClass`, less nuance. I think you did this for the paper mill?

        # clips ``{commodity}_out`` to ``min(uncurtailed, command_value)`` and
        # copies the raw output into ``uncurtailed_{commodity}_out``. It is a
        # no-op when no upstream controller is configured.
        self.apply_curtailment(outputs)
```

See `h2integrate/converters/solar/solar_pysam.py` for the full implementation,
including tilt-angle and resource-data handling.

```{note}
`setup` is where the configuration object is built and where any additional I/O
is registered. Always call `super().setup()` first so that the baseclass can
register the standard production outputs (and, for flexible models, the
command-value input). The `compute` signature is
`compute(self, inputs, outputs, discrete_inputs, discrete_outputs)` because
performance models may use discrete I/O (e.g. resource data dictionaries).
```

```{tip}
`merge_shared_inputs(model_inputs, kind)` combines
`model_inputs['{kind}_parameters']` and `model_inputs['shared_parameters']`
into a single dictionary, and raises if a key is defined in both. Pair it with
`BaseConfig.from_dict(..., strict=True)` so that unknown keys in `tech_config`
are flagged immediately.
```

> (required-class-attributes)=

Reviewer 2: It's kinda funky having a header nestled within the list...

For `flexible` models specifically, the baseclass automatically registers the `{commodity}_command_value` input and `uncurtailed_{commodity}_out` output, and the `compute()` method must call `self.apply_curtailment(outputs)` after writing the raw production to `outputs[f"{commodity}_out"]`. For `dispatchable` models the command value is consumed by the model's own internal logic; no curtailment helper is needed. `fixed` and `feedstock` models do not receive a command value at all.

3. **Write the cost model for your technology.**
Cost models follow the same pattern as performance models, but inherit from
`CostModelBaseClass` and use a `CostModelBaseConfig` (or `BaseConfig`)
configuration class. `CostModelBaseClass` registers the required `CapEx`,

> `OpEx`, `VarOpEx`, and `cost_year` outputs; no inputs are predefined.

Reviewer 2: `VarOpEx` has a default of zero so it's not a required output of a cost model persay

> ### Adding a new technology type

Reviewer 2: I feel like this is redundant with what's already in the adding a new technology section, but this seems like it's a little more concise and maybe better?
